In [35]:
import pandas as pd

In [36]:
data = pd.read_csv("./machine_learning_data/bank_data.csv",encoding='gbk')

### 查看数据前十行

In [37]:
data.head(10)

Unnamed: 0   custid                          trade_no bank_card_no  \
0           5  2791858  20180507115231274000000023057383          卡号1   
1          10   534047  20180507121002192000000023073000          卡号1   
2          12  2849787  20180507125159718000000023114911          卡号1   
3          13  1809708  20180507121358683000000388283484          卡号1   
4          14  2499829  20180507115448545000000388205844          卡号1   
5          15   518072  20180507121233054000000388275132          卡号1   
6          16  1205125  20180507121931540000000388298915          卡号1   
7          18  1129897  20180507124659235000000023105807          卡号1   
8          20  2599411  20180507115855621000000388224458          卡号1   
9          26  1413051  20180504155156296000000021138084          卡号1   

   low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   
5                0.02                   0.98   
6                0.02                   0.98   
7                0.02                   0.98   
8                0.03                   0.65   
9                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   
5                                   2000                               7.59   
6                                      0                              23.67   
7                                      0                               0.25   
8                                      0                               0.31   
9                                    500                               0.80   

   trans_activity_month  trans_activity_day        ...         \
0                  0.55               0.313        ...          
1                  1.00               0.458        ...          
2                  1.00               0.114        ...          
3                  0.57               0.777        ...          
4                  1.00               0.175        ...          
5                  1.00               0.733        ...          
6                  0.94               0.087        ...          
7                  0.88               0.302        ...          
8                  0.76               0.472        ...          
9                  1.00               0.088        ...          

   loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0           2900.0           1688.0                1200.0   
1           3500.0           1758.0               15100.0   
2           1600.0           1250.0                4200.0   
3           3200.0           1541.0               16300.0   
4           2300.0           1630.0                8300.0   
5           5300.0           1941.0               11200.0   
6           2200.0           2200.0                7600.0   
7              NaN              NaN                   NaN   
8           5300.0           4750.0                5500.0   
9           2800.0           1520.0                   0.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0          

### 查看数据集属性

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4754 entries, 0 to 4753
Data columns (total 90 columns):
Unnamed: 0                                    4754 non-null int64
custid                                        4754 non-null int64
trade_no                                      4754 non-null object
bank_card_no                                  4754 non-null object
low_volume_percent                            4752 non-null float64
middle_volume_percent                         4752 non-null float64
take_amount_in_later_12_month_highest         4754 non-null int64
trans_amount_increase_rate_lately             4751 non-null float64
trans_activity_month                          4752 non-null float64
trans_activity_day                            4752 non-null float64
transd_mcc                                    4752 non-null float64
trans_days_interval_filter                    4746 non-null float64
trans_days_interval                           4752 non-null float64
regional_mobility

In [39]:
# 查看数据类型已经对应类型个数
print(data.dtypes.value_counts())

float64    70
int64      13
object      7
dtype: int64


#### 查看数据集描述性统计

In [40]:
data.describe()

Unnamed: 0        custid  low_volume_percent  middle_volume_percent  \
count   4754.000000  4.754000e+03         4752.000000            4752.000000   
mean    6008.414178  1.690993e+06            0.021806               0.901294   
std     3452.071428  1.034235e+06            0.041527               0.144856   
min        5.000000  1.140000e+02            0.000000               0.000000   
25%     3106.000000  7.593358e+05            0.010000               0.880000   
50%     6006.500000  1.634942e+06            0.010000               0.960000   
75%     8999.000000  2.597905e+06            0.020000               0.990000   
max    11992.000000  4.004694e+06            1.000000               1.000000   

       take_amount_in_later_12_month_highest  \
count                            4754.000000   
mean                             1940.197728   
std                              3923.971494   
min                                 0.000000   
25%                                 0.000000   
50%                               500.000000   
75%                              2000.000000   
max                             68000.000000   

       trans_amount_increase_rate_lately  trans_activity_month  \
count                        4751.000000           4752.000000   
mean                           14.160674              0.804411   
std                           694.180473              0.196920   
min                             0.000000              0.120000   
25%                             0.615000              0.670000   
50%                             0.970000              0.860000   
75%                             1.600000              1.000000   
max                         47596.740000              1.000000   

       trans_activity_day   transd_mcc  trans_days_interval_filter  \
count         4752.000000  4752.000000                 4746.000000   
mean             0.365425    17.502946                   29.029920   
std              0.170196     4.475616                   22.722432   
min              0.033000     2.000000                    0.000000   
25%              0.233000    15.000000                   16.000000   
50%              0.350000    17.000000                   23.000000   
75%              0.480000    20.000000                   32.000000   
max              0.941000    42.000000                  285.000000   

             ...         loans_max_limit  loans_avg_limit  \
count        ...             4457.000000      4457.000000   
mean         ...             3390.038142      1820.357864   
std          ...             1474.206546       583.418291   
min          ...                0.000000         0.000000   
25%          ...             2300.000000      1535.000000   
50%          ...             3100.000000      1810.000000   
75%          ...             4300.000000      2100.000000   
max          ...            10000.000000      6900.000000   

       consfin_credit_limit  consfin_credibility  consfin_org_count_current  \
count           4457.000000          4457.000000                4457.000000   
mean            9187.009199            76.042630                   4.732331   
std             7371.257043            14.536819                   2.974596   
min                0.000000             0.000000                   0.000000   
25%             4800.000000            77.000000                   2.000000   
50%             7700.000000            79.000000                   4.000000   
75%            11700.000000            80.000000                   7.000000   
max            87100.000000            87.000000                  18.000000   

       consfin_product_count  consfin_max_limit  consfin_avg_limit  \
count            4457.000000        4457.000000        4457.000000   
mean                5.227507       16153.690823        8007.696881   
std                 3.409292       14301.037628        5679.418585   
min                 0.000000           0.000000           0.000000   
25%              

#### 查看每个特征的属性值分布情况

In [41]:
for col in data:
    print("{}:{}\r\n".format(col,data[col].value_counts()))

Unnamed: 0:2047     1
6774     1
8825     1
637      1
4735     1
10880    1
8833     1
6786     1
645      1
9143     1
4743     1
10888    1
8841     1
4747     1
10892    1
2704     1
6802     1
2680     1
8821     1
8807     1
6770     1
10902    1
605      1
6750     1
10848    1
609      1
4707     1
8805     1
6758     1
2664     1
        ..
3371     1
11567    1
9178     1
7473     1
1330     1
3379     1
5428     1
7477     1
1334     1
3383     1
7481     1
7461     1
5412     1
11444    1
5392     1
3331     1
3335     1
1290     1
9568     1
9486     1
10317    1
11308    1
3359     1
10945    1
3347     1
7445     1
1302     1
3355     1
7453     1
8192     1
Name: Unnamed: 0, Length: 4754, dtype: int64

custid:284669     1
1077931    1
309934     1
1045167    1
139340     1
2396855    1
1403580    1
2071230    1
2620098    1
1569475    1
1837764    1
328389     1
2411208    1
631501     1
2661070    1
2609871    1
2665168    1
815491     1
1881943    1
2691859    1
56745

consfin_org_count_current:3.0     588
4.0     580
2.0     549
5.0     511
1.0     465
6.0     447
7.0     351
8.0     288
9.0     198
0.0     151
10.0    144
11.0     88
12.0     45
13.0     25
14.0     20
15.0      3
16.0      2
18.0      1
17.0      1
Name: consfin_org_count_current, dtype: int64

consfin_product_count:3.0     530
4.0     510
2.0     492
5.0     470
1.0     447
6.0     406
7.0     372
8.0     304
9.0     226
10.0    181
0.0     151
11.0    137
12.0     96
13.0     56
14.0     40
15.0     18
16.0     11
17.0      5
18.0      3
20.0      2
Name: consfin_product_count, dtype: int64

consfin_max_limit:20400.0     257
19200.0     206
14400.0     197
8400.0      190
13200.0     185
7200.0      162
0.0         151
7800.0      150
13800.0     141
18000.0     134
4800.0      126
15600.0     126
9600.0      119
22800.0     118
10800.0     112
16800.0     109
21600.0     106
12000.0     100
19800.0      87
3600.0       86
6000.0       78
26400.0      60
2400.0       60
1000.0  

#### 通过上述代码结果分析发现:
    无关特征：unnamed:0、custid、trade_no、bank_card_no、id_name
    特殊特征：student_feature、reg_preference_for_trad、source
    日期特征：latest_query_time、loans_latest_time

In [42]:
# 打印以上特征属性值分布情况
special_features = ['Unnamed: 0', 'custid','bank_card_no','id_name',
                    'student_feature','reg_preference_for_trad','source','lasest_query_time','loans_lastest_time']
for col in special_features:
    print('{}:{}\r\n'.format(col, data[col].value_counts()))

Unnamed: 0:2047     1
6774     1
8825     1
637      1
4735     1
10880    1
8833     1
6786     1
645      1
9143     1
4743     1
10888    1
8841     1
4747     1
10892    1
2704     1
6802     1
2680     1
8821     1
8807     1
6770     1
10902    1
605      1
6750     1
10848    1
609      1
4707     1
8805     1
6758     1
2664     1
        ..
3371     1
11567    1
9178     1
7473     1
1330     1
3379     1
5428     1
7477     1
1334     1
3383     1
7481     1
7461     1
5412     1
11444    1
5392     1
3331     1
3335     1
1290     1
9568     1
9486     1
10317    1
11308    1
3359     1
10945    1
3347     1
7445     1
1302     1
3355     1
7453     1
8192     1
Name: Unnamed: 0, Length: 4754, dtype: int64

custid:284669     1
1077931    1
309934     1
1045167    1
139340     1
2396855    1
1403580    1
2071230    1
2620098    1
1569475    1
1837764    1
328389     1
2411208    1
631501     1
2661070    1
2609871    1
2665168    1
815491     1
1881943    1
2691859    1
56745

KeyError: 'lasest_query_time'

### 初步结论
(1) 数据集共有4754个样本，特征维度为90，其中float64类型为70个，int64类型为13个，object类型为7个 <br>
<br>
(2) 通过info发现，数据集除了'student_feature'存在大量缺失值外，在整体表现是相当完整的 <br>
<br>
(3) 大部分特征的方差很大，存在拖尾现象<br>
<br>
(4) 初步分析得到的无关特征有['Unnamed: 0','custid','trade_no','bank_card_no','id_name','student_feature']、需要进行类型转换的特征有：['reg_preference_for_trad']、日期特征有：['source','latest_query_time','loans_latest_time']
<br><br>
(5) 后续还需要通过寻找特征相关性和特征组合对特征进行选择和提取

### 无关特征删除

In [43]:
#对上述分析得到的无关特征['Unnamed: 0','custid','trade_no','bank_card_no','id_name','student_feature']+日期特征进行删除
drop_features = ['Unnamed: 0','custid','trade_no','bank_card_no','id_name',
                    'student_feature','source','latest_query_time','loans_latest_time']
data_del_useless = data.drop(drop_features,axis=1)
print(data_del_useless.dtypes.value_counts())

float64    69
int64      11
object      1
dtype: int64


### 数据类型转换

#### 将城市等级：'reg_preference_for_trad'进行类型转换

In [44]:
# 查看'reg_preference_for_trad'的所有属性
data_del_useless['reg_preference_for_trad'].value_counts()

一线城市    3403
三线城市    1064
境外       150
二线城市     131
其他城市       4
Name: reg_preference_for_trad, dtype: int64

In [46]:

for i in range(0, data.shape[0]):
    if data_del_useless.reg_preference_for_trad[i] == '一线城市':
        data_del_useless.reg_preference_for_trad[i] = 1
    elif data_del_useless.reg_preference_for_trad[i] == '二线城市':
        data_del_useless.reg_preference_for_trad[i] = 2
    elif data_del_useless.reg_preference_for_trad[i] == '三线城市':
        data_del_useless.reg_preference_for_trad[i] = 3
    elif data_del_useless.reg_preference_for_trad[i] == '境外':
        data_del_useless.reg_preference_for_trad[i] = 4
    elif data_del_useless.reg_preference_for_trad[i] == '其他城市':
        data_del_useless.reg_preference_for_trad[i] = 5
data_trans = data_del_useless.apply(pd.to_numeric, errors="ignore")
print(data_trans['reg_preference_for_trad'].value_counts())
print(data_trans.dtypes.value_counts())

1.0    3403
3.0    1064
4.0     150
2.0     131
5.0       4
Name: reg_preference_for_trad, dtype: int64
float64    70
int64      11
dtype: int64


### 缺失值处理
(1) 删除缺失值超过30%的行或列
<br>
(2) 将缺失值用每列的mean代替

In [47]:
# 删除缺失值超过30%的行
data_procna = data_trans.dropna(axis=0,thresh=data_trans.shape[1]*0.7)
# 删除缺失值超过30%的列
data_procna = data_procna.dropna(axis=1,thresh=data_trans.shape[0]*0.7)
# 使用均值代替剩下的缺失值
data_procna = data_procna.fillna(data_procna.mean())

In [48]:
data_procna

low_volume_percent  middle_volume_percent  \
0                   0.01                   0.99   
1                   0.02                   0.94   
2                   0.04                   0.96   
3                   0.00                   0.96   
4                   0.01                   0.99   
5                   0.02                   0.98   
6                   0.02                   0.98   
8                   0.03                   0.65   
9                   0.01                   0.99   
10                  0.01                   0.99   
11                  0.01                   0.99   
12                  0.00                   0.73   
13                  0.01                   0.79   
14                  0.01                   0.89   
15                  0.01                   0.99   
16                  0.02                   0.98   
17                  0.02                   0.98   
18                  0.01                   0.99   
19                  0.01                   0.96   
20                  0.02                   0.98   
21                  0.02                   0.98   
22                  0.00                   0.59   
23                  0.01                   0.99   
24                  0.02                   0.78   
25                  0.40                   0.60   
26                  0.02                   0.96   
27                  0.01                   0.91   
28                  0.01                   0.99   
29                  0.02                   0.64   
30                  0.02                   0.93   
...                  ...                    ...   
4723                0.01                   0.96   
4724                0.02                   0.98   
4725                0.01                   0.99   
4726                0.00                   1.00   
4727                0.02                   0.98   
4728                0.06                   0.94   
4729                0.17                   0.83   
4730                0.01                   0.99   
4731                0.00                   0.58   
4732                0.11                   0.89   
4733                0.03                   0.97   
4734                0.00                   1.00   
4735                0.01                   0.61   
4736                0.05                   0.95   
4737                0.01                   0.99   
4738                0.02                   0.98   
4739                0.05                   0.62   
4740                0.07                   0.93   
4741                0.00                   0.83   
4742                0.02                   0.98   
4743                0.00                   1.00   
4744                0.01                   0.96   
4745                0.04                   0.96   
4747                0.02                   0.40   
4748                0.09                   0.91   
4749                0.05                   0.68   
4750                0.01                   0.99   
4751                0.00                   1.00   
4752                0.05                   0.95   
4753                0.01                   0.99   

      take_amount_in_later_12_month_highest  \
0                                         0   
1                                      2000   
2                                         0   
3                                      2000   
4                                         0   
5                                      2000   
6                                         0   
8                                         0   
9                                       500   
10                                        0   
11                                      500   
12                                     5000   
13                                        0   
14                                     4000   
15                                        0   
16                                        0   
17                                      300 

### 寻找相关性

In [51]:
corr_matrix = data_procna.corr()
corr_matrix['status'].sort_values(ascending=False)

status                                        1.000000
trans_fail_top_count_enum_last_1_month        0.330420
history_fail_fee                              0.320066
loans_overdue_count                           0.279232
latest_one_month_fail                         0.257207
rank_trad_1_month                             0.147077
trans_fail_top_count_enum_last_6_month        0.130946
top_trans_count_last_1_month                  0.116782
trans_fail_top_count_enum_last_12_month       0.116548
trans_day_last_12_month                       0.074017
avg_price_top_last_12_valid_month             0.069930
latest_six_month_loan                         0.062971
latest_query_day                              0.059355
query_cash_count                              0.054434
latest_three_month_loan                       0.052953
low_volume_percent                            0.051755
loans_count                                   0.048461
first_transaction_time                        0.040481
middle_vol

### 数据切分

In [59]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(data_procna, test_size=0.3, random_state=2018)
train_data.to_csv('./machine_learning_data/bank_train_data.csv', index=False, header=True)
test_data.to_csv('./machine_learning_data/bank_test_data.csv', index=False, header=True)